core/server/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -170,89 +187,251 @@ public boolean inodeIdExists(long id) {
   public boolean inodePathExists(AlluxioURI path) {
     try {
       TraversalResult traversalResult =
-          traverseToInode(PathUtils.getPathComponents(path.toString()), false);
+          traverseToInode(PathUtils.getPathComponents(path.toString()), LockMode.READ, null);
+      traversalResult.getInodeLockGroup().close();
       return traversalResult.isFound();
     } catch (InvalidPathException e) {
       return false;
     }
   }
 
   /**
-   * @param id the id to get the inode for
-   * @return the inode with the given id
-   * @throws FileDoesNotExistException if the file does not exist
+   * Locks existing inodes on the specified path, in the specified {@link LockMode}. The target
+   * inode is not required to exist.
+   *
+   * @param path the path to lock
+   * @param lockMode the {@link LockMode} to lock the inodes with
+   * @return the {@link InodePath} representing the locked path of inodes
+   * @throws InvalidPathException if the path is invalid
    */
-  public Inode<?> getInodeById(long id) throws FileDoesNotExistException {
-    Inode<?> inode = mInodes.getFirstByField(mIdIndex, id);
-    if (inode == null) {
-      throw new FileDoesNotExistException("Inode id " + id + " does not exist.");
+  public InodePath lockInodePath(AlluxioURI path, LockMode lockMode) throws InvalidPathException {
+    TraversalResult traversalResult =
+        traverseToInode(PathUtils.getPathComponents(path.getPath()), lockMode, null);
+    return new ExtensibleInodePath(path, traversalResult.getInodes(),
+        traversalResult.getInodeLockGroup());
+  }
+
+  /**
+   * Locks existing inodes on the two specified paths. The two paths will be locked in the
+   * correct order. The target inodes are not required to exist.
+   *
+   * @param path1 the first path to lock
+   * @param lockMode1 the {@link LockMode} of the first path
+   * @param path2 the second path to lock
+   * @param lockMode2 the {@link LockMode} of the second path
+   * @return a {@link InodePathPair} representing the two locked paths
+   * @throws InvalidPathException if a path is invalid
+   */
+  public InodePathPair lockInodePathPair(AlluxioURI path1, LockMode lockMode1, AlluxioURI path2,
+      LockMode lockMode2) throws InvalidPathException {
+    String[] pathComponents1 = PathUtils.getPathComponents(path1.getPath());
+    String[] pathComponents2 = PathUtils.getPathComponents(path2.getPath());
+    List<LockMode> lockHints = new ArrayList<>();
+
+    int minLength = Math.min(pathComponents1.length, pathComponents2.length);
+    for (int i = 0; i < minLength; i++) {
+      if (pathComponents1[i].equals(pathComponents2[i])) {
+        // The two paths share a common path prefix.
+        LockMode mode1 = getLockModeForComponent(i, pathComponents1.length, lockMode1, null);
+        LockMode mode2 = getLockModeForComponent(i, pathComponents2.length, lockMode1, null);
+        // If either of the modes are WRITE, lock both paths as WRITE to prevent deadlock.
```

Comment:
`getLockModeForComponent` only returns READ or WRITE, for each component of the path. I added a todo for a combiner, since I think this is the only place we would use it.